adds see more button #1295

mauditecandela opened the pull request:
- adds button to revisions_handler and onClick method
- adds translation to english for the message in the button
- includes a limit param to the courses controller
- creates new server_action and api functions to generate the endpoints

to do:
- show the new limit in the UI

mauditecandela commented:
@ragesoss this is my first commit to the topic. Afaik, it works from server side perspective but not in the frontend. I will check now how to serve the new limit of rows. meanwhile, I thought it could be nice if you could give me some feedback, specially on creating new methods for the server_actions and api. Thanks a lot!

ragesoss commented on app/assets/javascripts/actions/server_actions.js:
```diff
@@ -13,6 +13,14 @@ const ServerActions = Flux.createActions({
   },
 
+  fetchMoreRevisions(courseId, limit) {
+    const actionType = `RECEIVE_MORE_REVISIONS`;
```
I think the disconnection with rendering the revisions is here. The action type determines which stores will receive and process the fetched data, and there is no store that handles `RECEIVE_MORE_REVISIONS`. If you change it to `RECEIVE_REVISIONS`, that may fix it.

I would suggest, rather than 'fetchMoreRevsions`, you just call this `fetchRevisions`, and use the same action (just with the default limit) in componentWillMount.

mauditecandela replied:
What do you exactly mean? 'fetchRevisions' already exist and I think it does retrieve different endpoint (revisions for each student). I have tried and didn't work, but I will check it further. 
I added though 'RECEIVE_REVISIONS' and indeed works. Thanks.

ragesoss replied:
You're right, I misread the code. Currently, the default fetch revisions for the course is going through API.fetch(courseId, model). So the way to avoid using two different methods for basically the same thing, you could call the new method `fetchCourseRevisions`, and use it instead of general-purpose fetch method for both the default and 'more' data flows.

ragesoss commented on app/assets/javascripts/utils/api.js:
```diff
@@ -66,6 +66,23 @@ const API = {
     });
   },
 
+  fetchMoreRevisions(courseId, limit) {
```
Similarly, here you can probably just adjust fetchRevisions to take a limit, and then stick with a single fetch method here.

ragesoss commented on app/controllers/courses_controller.rb:
```diff
@@ -215,6 +216,10 @@ def set_endpoint
   end
 
+  def set_limit
+    @limit = (params[:limit] || 50)
```
This points to a problem with how we've set up the multipurpose `show` endpoint for courses. The limit is relevant only for a single endpoint, but this will set it for every endpoint. In the short term, another option might be to handle the limit param inside the jbuilder view.

In the longer term, I think we need to rewrite the routing and controller to have an explicit endpoint for each type of json data for a course.

mauditecandela replied:
I limited the set_limit to the revisions endpoint with an IF statement, as I think it's clearer if it's in the controller. Why do you think it's better to add it to the json jbuilder?

ragesoss replied:
I'm okay with the if statement. That should make it clear that this is only relevant for the one endpoint.

ragesoss commented on app/controllers/courses_controller.rb:
```diff
@@ -217,7 +217,7 @@ def set_endpoint
 
   def set_limit
-    @limit = (params[:limit] || 50)
+    @limit = (params[:limit] || 50) if params[:endpoint] = 'revisions'
```
Should be `==`

mauditecandela commented:
@ragesoss the tests should pass. 
I added states in the revision_handler in order to define the number of activities to show. I also changed the name of the fetching method + added to componentWillMount in order to have the same structure. Let me know what do you think! :)

ragesoss commented on app/assets/javascripts/components/revisions/revisions_handler.jsx:
```diff
@@ -8,17 +8,29 @@ const RevisionHandler = React.createClass({
 
   componentWillMount() {
-    return ServerActions.fetch('revisions', this.props.course_id);
+    this.setState({ limit: 100 });
```
This shouldn't be in componentWillMount, I think.

mauditecandela replied:
Do you mean that it's a mistake or that shouldn't be placed there?

ragesoss replied:
This hook runs immediately when the component starts to load, so it would overwrite the default of 50 from getInitialState. I think you also can't set state from this React lifecycle hook anyway.

mauditecandela replied:
I thought so but it works in my machine. I tried it with lower limit numbers and worked perfectly. componentWillMount was both taking the limit from the getInitialState and setting it on 100 for the next interaction. The same with the setState in showMore(), it worked as expected.

ragesoss replied:
I see we have setState a few places elsewhere within componentWillMount, so that seems to be okay technically. But it should be necessary as far as I can tell.

ragesoss commented on app/assets/javascripts/utils/api.js:
```diff
@@ -65,6 +65,23 @@ const API = {
     });
   },
+  
+  fetchMoreRevisions(courseId, limit) {
```
This should be named fetchCourseRevisions as well.

mauditecandela replied:
Done! 👍

mauditecandela commented:
I don't understand why the last commit could deliver to that error 😓 Do you think it's related?

ragesoss commented:
Nope, that's unrelated. It's a flaky feature spec that fails randomly on occasion. I restarted the build; it should pass this time, unless it gets unlucky twice in a row.

ragesoss commented on app/assets/javascripts/components/revisions/revisions_handler.jsx:
```diff
@@ -8,17 +8,28 @@ const RevisionHandler = React.createClass({
 
+  showMore() {
+    this.setState({ limit: this.state.limit + 100 });
+    return ServerActions.fetchCourseRevisions(this.props.course_id, (this.state.limit + 50));
```
I don't understand what is going on here. Why is setState adding 100, but +50 is used in the following line?

mauditecandela replied:
I agree this solution was rather confusing - tbh I was not that happy with it. I just commited a new one, let me know if you think it's better. :)

ragesoss replied:
Huh... I didn't think this would work, but it does... I guess it's possible to assign to this.state directly, even though it's considered "wrong": https://reactjs.org/docs/state-and-lifecycle.html#using-state-correctly

But I don't think we need to treat the first 'more' any differently from the rest. Here's what I suggest:

```
const newLimit = this.state.limit + 100;
this.setState({ limit: newLimit });
return ServerActions.fetchCourseRevisions(this.props.course_id, newLimit);
```

ragesoss commented:
@mauditecandela nice work! I'll go ahead a merge this. A further improvement would be to only show the button if the number of revisions at or close to the limit. That way, it won't appear a) while fetching of new revisions is happening, and b) when all the available revisions are already being shown.

mauditecandela commented:
@ragesoss Thanks for the hint! :) I was planning on having a look to issue #1005 but I will check what you are mentioning as well.

ragesoss commented:
@mauditecandela I opened a new issue, #1451, for it, but see my notes there... it might be trickier than it seems.
